Assignment of column via .loc for numpy non-ns datetimes

moon1ite opened the pull request:
- [x] closes #27395 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

moon1ite committed:
Fix issue 27395

moon1ite committed:
Fix lint error

jreback commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -32,7 +32,7 @@ Categorical
 ^^^^^^^^^^^^
 - Bug in :func:`to_datetime` where passing a timezone-naive :class:`DatetimeArray` or :class:`DatetimeIndex` and ``utc=True`` would incorrectly return a timezone-naive result (:issue:`27733`)
--
+- Bug in :func:`maybe_cast_to_datetime` where converting a `np.datetime64` to `datetime64[D]` raise `TypeError` (:issue: `27395`)
```
this is a private function, you would have to indicate the user facing experience here instead (e.g. the assignment)

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1026,7 +1026,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
 
             if is_datetime64 and not is_dtype_equal(dtype, _NS_DTYPE):
-                if dtype.name in ("datetime64", "datetime64[ns]"):
+                if dtype.name in ("datetime64", "datetime64[ns]", "datetime64[D]"):
```
just use ```dtype.kind == 'M'```

inmoonlight replied:
Since `pandas` doesn't support to convert [`datetime[ps]`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_constructors.py#L1352), I used `(dtype.kind == 'M') and (dtype.name != 'datetime64[ps]')`

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1044,7 +1044,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
 
             elif is_timedelta64 and not is_dtype_equal(dtype, _TD_DTYPE):
-                if dtype.name in ("timedelta64", "timedelta64[ns]"):
+                if dtype.name in ("timedelta64", "timedelta64[ns]", "timedelta64[D]"):
```
```dtype.kind == 'm'```

inmoonlight replied:
Since `pandas` doesn't support to convert [`datetime[ps]`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_constructors.py#L1352), I used `(dtype.kind == 'm') and (dtype.name != 'timedelta64[ps]')`

jreback commented on pandas/tests/dtypes/cast/test_infer_dtype.py:
```diff
@@ -169,3 +170,14 @@ def test_cast_scalar_to_array(obj, dtype):
     tm.assert_numpy_array_equal(arr, exp)
+
+
+@pytest.mark.parametrize(
```
its ok to test this i guess, but would need to parameterize on datetime64[D], ns, datetime64 etc and check the result

inmoonlight replied:
done. I made all possible datetime formats except [ps] (as mentioned above)

TomAugspurger replied:
I think @jreback's comment was requesting testing those other frequencies here.

And then assert the result matches too.

```python
result = maybe_cast_to_datetime(obj, dtype)
expected = pd.Timestamp(...)  # right?
assert result == expected
```

jreback reviewed (changes requested):
the most important types of tests are replicating the OP indexing issue. put in pandas/tests/indexing/test_loc.py (see the setting tests)

moon1ite committed:
shoot numpy-dev future warning

moon1ite committed:
Reflect reviews

- update doc
- update dtype matching fucntion
- update test

moon1ite committed:
Fix ci error

moon1ite committed:
Fix dtype match except [ps]

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -85,6 +84,7 @@ Indexing
 - Bug in partial-string indexing returning a NumPy array rather than a ``Series`` when indexing with a scalar like ``.loc['2015']`` (:issue:`27516`)
 - Break reference cycle involving :class:`Index` and other index classes to allow garbage collection of index objects without running the GC. (:issue:`27585`, :issue:`27840`)
 - Fix regression in assigning values to a single column of a DataFrame with a ``MultiIndex`` columns (:issue:`27841`).
+- Fix assignment of column via `.loc` with numpy `non-ns datetime` type (:issue: `27395`)
```
```suggestion
- Fix assignment of column via `.loc` with numpy `non-ns datetime` type (:issue:`27395`)
```

I'm not sure if sphinx will complain about the space or not.

Also, I think the backticks on "non-ns datetime" aren't necessary.

TomAugspurger commented on pandas/tests/indexing/test_loc.py:
```diff
@@ -690,6 +690,32 @@ def test_loc_setitem_consistency_slice_column_len(self):
+            }
+        )
+        df.loc[:, "day"] = df.loc[:, "timestamp"].values.astype(dtype)
+
```
Can you add an assert about the result here? `tm.assert_series_equal(df['day'], expected)`

inmoonlight replied:
While I update the tests, I found a strange result. (Not sure this is intended.)

```python
df = pd.DataFrame({'timestamp': [np.datetime64('2017-01-01 01:02:23'), np.datetime64('2018-11-12 12:34:12')]})
df['year'] = df.loc[:, 'timestamp'].values.astype('datetime64[Y]')
df['month'] = df.loc[:, 'timestamp'].values.astype('datetime64[M]')
df.loc[:, 'day'] = df.loc[:, 'timestamp'].values.astype('datetime64[D]')
```
result:
```
            timestamp        day      month       year
0 2017-01-01 01:02:23 2017-01-01 2017-01-01 2017-01-01
1 2018-11-12 12:34:12 2018-11-12 2018-11-01 2018-01-01
```
`astype` result is okay.

```python
df.loc[:, 'timestamp'].values.astype('datetime64[Y]')
```
result:
```
array(['2017', '2018'], dtype='datetime64[Y]')
```

jbrockmendel replied:
Without running the test in master, it isn't obvious to the reader what is being tested.  I suggest something like:

```
@pytest.mark.parametrize("unit", ...)
def test_loc_setitem_datetime64_non_ns_converts(self, unit):
     # GH#27928 a few words about how this used to break
    df = DataFrame(
            {
                "timestamp": [
                    np.datetime64("2017-02-11 12:41:29"),
                    np.datetime64("1991-11-07 04:22:37"),
                ]
            }
        )
        df[unit] = df["timestamp"].values.astype("datetime64[{unit}]".format(unit=unit))
        df.loc[:, unit] = df["timestamp"].values.astype("datetime64[{unit}]".format(unit=unit))
```

TomAugspurger reviewed (commented):
Looks close. A few questions on the tests.

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1055,7 +1055,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
 
             if is_scalar(value):
-                if value == iNaT or isna(value):
+                if value is iNaT or isna(value):
```
@jbrockmendel something to address later, we need to separate integer tests here

@inmoonlight is there a reason you changed this?

jbrockmendel replied:
The integer iNaT is not interned, so this seems dangerous

inmoonlight replied:
@jreback 
I found this led to the failure in `pandas-dev.pandas (Linux py37_np_dev)` env. If you all think it is dangerous, I'll put it back.

TomAugspurger replied:
Restore the original I think. We've had some flaky tests in master recently that we're fixing elsewhere.

inmoonlight replied:
@TomAugspurger 
Restored it.

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1044,7 +1044,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
 
             elif is_timedelta64 and not is_dtype_equal(dtype, _TD_DTYPE):
-                if dtype.name in ("timedelta64", "timedelta64[ns]"):
+                if (dtype.kind == "m") and (dtype.name != "timedelta64[ps]"):
```
let's leave [ps] for another issue. I am not sure what we should do about it. we *could* convert it with precision loss rather than actually raising (which is what I think it would do now)

inmoonlight replied:
@jreback 
Without this, ci check failed... Does this imply that my modification is wrong?

TomAugspurger replied:
Do you remember the CI failure? It may have been a flaky test.

inmoonlight replied:
@TomAugspurger 
Yes, the failure part is: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_constructors.py#L1352

inmoonlight replied:
@jreback 

It does not raise an error and convert [ps] resolution, but,,, the value goes wrong.

```python
df = pd.DataFrame({'timestamp': [np.datetime64('2017-01-01 01:00:00'), np.datetime64('2017-01-01 02:00:00')],
                   'value': ['a', 'b']})
df['ps'] = df.loc[:, 'timestamp'].values.astype('datetime64[ps]')
```
```
            timestamp value                            ps
0 2017-01-01 01:00:00     a 1972-03-20 01:55:03.235870720
1 2017-01-01 02:00:00     b 1972-04-30 17:55:03.235870720
```

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into inmoonlight-hotfix/27395/loc-assignment

moon1ite committed:
Reflect reviews

- update docs
- remove previous tests
- update tests
- revert to the changes that made to pass the ci tests

moon1ite committed:
Fix conflicts

TomAugspurger commented:
Looks like there are still some CI failures. Do you need help looking into
them?

On Wed, Aug 21, 2019 at 10:25 AM Jihyung Moon <notifications@github.com>
wrote:

> *@inmoonlight* commented on this pull request.
> ------------------------------
>
> In pandas/core/dtypes/cast.py
> <https://github.com/pandas-dev/pandas/pull/27928#discussion_r316249730>:
>
> > @@ -1055,7 +1055,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
>                      )
>
>              if is_scalar(value):
> -                if value == iNaT or isna(value):
> +                if value is iNaT or isna(value):
>
> @TomAugspurger <https://github.com/TomAugspurger>
> Restored it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/27928?email_source=notifications&email_token=AAKAOITQBERVCNCF3RMLKI3QFVM7XA5CNFSM4IL3XFI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCIBNBQ#discussion_r316249730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOISVG63DGWO27U3CKWTQFVM7XANCNFSM4IL3XFIQ>
> .
>

moon1ite committed:
Fix CI error

inmoonlight commented:
@TomAugspurger 

Desperately...

I looked into several CI issues and found they all raise the same error.

```
=================================== FAILURES ===================================
_ TestSeriesConstructors.test_constructor_generic_timestamp_bad_frequency[m8[ps]-cannot convert timedeltalike] _
[gw0] linux -- Python 3.7.4 /home/travis/miniconda3/envs/pandas-dev/bin/python
self = <pandas.tests.series.test_constructors.TestSeriesConstructors object at 0x7f2b6f5c2910>
dtype = 'm8[ps]', msg = 'cannot convert timedeltalike'
    @pytest.mark.parametrize(
        "dtype,msg",
        [
            ("m8[ps]", "cannot convert timedeltalike"),
            ("M8[ps]", "cannot convert datetimelike"),
        ],
    )
    def test_constructor_generic_timestamp_bad_frequency(self, dtype, msg):
        # see gh-15524, gh-15987
    
        with pytest.raises(TypeError, match=msg):
>           Series([], dtype=dtype)
E           Failed: DID NOT RAISE <class 'TypeError'>
pandas/tests/series/test_constructors.py:1363: Failed
_ TestSeriesConstructors.test_constructor_generic_timestamp_bad_frequency[M8[ps]-cannot convert datetimelike] _
[gw0] linux -- Python 3.7.4 /home/travis/miniconda3/envs/pandas-dev/bin/python
self = <pandas.tests.series.test_constructors.TestSeriesConstructors object at 0x7f2b6f5df750>
dtype = 'M8[ps]', msg = 'cannot convert datetimelike'
    @pytest.mark.parametrize(
        "dtype,msg",
        [
            ("m8[ps]", "cannot convert timedeltalike"),
            ("M8[ps]", "cannot convert datetimelike"),
        ],
    )
    def test_constructor_generic_timestamp_bad_frequency(self, dtype, msg):
        # see gh-15524, gh-15987
    
        with pytest.raises(TypeError, match=msg):
>           Series([], dtype=dtype)
E           Failed: DID NOT RAISE <class 'TypeError'>
pandas/tests/series/test_constructors.py:1363: Failed
``` 

At first glance, I thought pandas does not support [ps] level. This led to https://github.com/pandas-dev/pandas/pull/27928/commits/98a5061bbcc04bb6114cbf1bf69efe15f0e66432 .

How could I solve it?

TomAugspurger commented:
> At first glance, I thought pandas does not support [ps] level.

That's correct. I think if the user explicitly requests `[ps]` we raise, and if we're giving a `[ps]`-resolution array we convert to `[ns]` (is that correct @jbrockmendel?)

jbrockmendel commented:
I think that's right.  xref #23571

moon1ite committed:
Fix conficts

moon1ite committed:
Fix CI error

moon1ite committed:
Fix CI error

inmoonlight commented:
@TomAugspurger 

So,,, raising ValueError when allocating [ps] resolution is an issue to be solved...
Which means `pandas/tests/series/test_constructors.py:1363` can be removed?

Am I right? 🤔

TomAugspurger commented:
> Which means pandas/tests/series/test_constructors.py:1363 can be removed?

IIUC, I think that should stay. I think these are the expected behaviors

```python
>>> pd.Series([], dtype='datetime64[ps]')                                    # raises
>>> pd.Series(np.array([], dtype='datetime64[ps]'), dtype='datetime64[ps]')  # raises
>>> pd.Series(np.array([], dtype='datetime64[ps]'))                          # OK
```

inmoonlight commented:
@TomAugspurger 

Thanks for the explanations!
The final modification raises an error on those inputs.
Still more modification required?

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -1026,7 +1026,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
 
             if is_datetime64 and not is_dtype_equal(dtype, _NS_DTYPE):
-                if dtype.name in ("datetime64", "datetime64[ns]"):
+                if dtype <= np.dtype("M8[ns]"):
```
What does `<=` do for numpy dtypes? Check if it's a subtype?

jbrockmendel replied:
Not clear to me either:

```
>>> "timedelta64[ns]" < np.dtype("m8[ns]")
False
>>> "timedelta64[us]" < np.dtype("m8[ns]")
True
>>> "timedelta64[ps]" > np.dtype("m8[ns]")
True
>>> "timedelta64" < np.dtype("m8[ns]")
True
```

TomAugspurger commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -49,7 +49,7 @@ Interval
 ^^^^^^^^
 
--
+- Fix assignment of column via `.loc` with numpy non-ns datetime type (:issue:`27395`)
```
This will need to go to v1.0.0.rst now, sorry.

TomAugspurger reviewed (commented):
@jbrockmendel can you take another look here?

moon1ite committed:
Merge branch 'master' into hotfix/27395/loc-assignment

moon1ite committed:
Update whatsnew

moon1ite committed:
Add comments

moon1ite committed:
Merge branch 'master' into hotfix/27395/loc-assignment

moon1ite committed:
Fix CI error

moon1ite committed:
Fix CI error

jreback commented:
thanks @inmoonlight
